File: 52.n皇后-ii.cpp
/*
 * @lc app=leetcode.cn id=52 lang=cpp
 *
 * [52] N皇后 II
 */

// @lc code=start
#include <iostream>
#include <vector>
#include <cmath>

using namespace std;

class Solution {
public:
    int res = 0;
    bool pass(vector<vector<bool>> &bds, int row, int col, int n) {

        int vsum = 0;
        int dsum = 0;
        int ddsum = 0;

        for(int i = row; i >-1; --i) {
            vsum += bds[i][col];
        }

        int j = (row<col)?row:col;
        for(int i = 0; i <= j; ++i) {
            dsum += bds[row-i][col-i];
        }

        j = (row<(n-col))?row:(n-col);
        for(int i =0; i <= j; ++i) {
            ddsum += bds[row-i][col+i];
        }

        if (vsum > 1 || dsum > 1 || ddsum > 1)
            return false;
        return true;
    }

    void check(vector<vector<bool>> &bds, int row, int n) {

        for(int i = 0; i < n; ++i) {
            bds[row][i] = true;
            if(pass(bds, row, i, n)) {
                if(row < n-1)
                    check(bds, row+1, n);
                else{
                    this->res++;
                } 
            }
            bds[row][i] = false;
        }
    }

    int totalNQueens(int n) {
        vector<vector<bool>> bds;
        for(int i = 0; i < n; ++i){
            vector<bool> bd(n, false);
            bds.emplace_back(bd);
        }

        this->res = 0;
        check(bds, 0, n);
        bds.clear();
        return this->res;
    }
};

int main() {
    
    for(int n = 0; n < 10; ++n) {
        Solution s = Solution();
        int res = s.totalNQueens(n);
        cout << "NQueens(" << n << ")=" << res << endl;
        
    }

}
// @lc code=end

File: 52.n皇后-ii.py
class Solution:
    def __init__(self):
        res = 0

    def pass_(self, bds, row, col, n):
        vsum = 0
        dsum = 0
        ddsum = 0

        for i in range(row, -1, -1):
            vsum += bds[i][col]
        
        j = row if row < col else col
        for i in range(0, j+1):
            dsum += bds[row-i][col-i]
        
        j = row if row < (n-col-1) else (n-col-1)
        for i in range(0, j+1):
            ddsum += bds[row-i][col+i]

        if vsum > 1 or dsum > 1 or ddsum > 1:
            return False
        return True

    def check(self, bds, row, n):
        for i in range(n):
            bds[row][i] = 1
            if self.pass_(bds, row, i, n):
                if row < n-1:
                    self.check(bds, row+1, n)
                else:
                    self.res += 1
            bds[row][i] = 0

    def totalNQueens(self, n: int) -> int:
        self.res = 0
        bds = [[0 for i in range(n)] for _ in range(n)]
        self.check(bds, 0, n)
        return self.res
        

if __name__ == "__main__":
    s = Solution()
    a = []
    for i in range(13):
        res = s.totalNQueens(i)
        # print(f'NQueens({i}) = {res}')
        a.append(res)
    print(a)